site/src/modules/workspaces/WorkspaceTiming/WorkspaceTimings.tsx

```diff
@@ -4,44 +4,70 @@ import KeyboardArrowUp from "@mui/icons-material/KeyboardArrowUp";
 import Button from "@mui/material/Button";
 import Collapse from "@mui/material/Collapse";
 import Skeleton from "@mui/material/Skeleton";
-import type { AgentScriptTiming, ProvisionerTiming } from "api/typesGenerated";
+import type {
+	AgentConnectionTiming,
+	AgentScriptTiming,
+	ProvisionerTiming,
+} from "api/typesGenerated";
 import { type FC, useState } from "react";
 import { type TimeRange, calcDuration, mergeTimeRanges } from "./Chart/utils";
 import { ResourcesChart, isCoderResource } from "./ResourcesChart";
 import { ScriptsChart } from "./ScriptsChart";
-import { type StageCategory, StagesChart, stages } from "./StagesChart";
+import {
+	type Stage,
+	StagesChart,
+	agentStages,
+	provisioningStages,
+} from "./StagesChart";
 
 type TimingView =
 	| { name: "default" }
 	| {
 			name: "detailed";
-			stage: string;
-			category: StageCategory;
+			stage: Stage;
 			filter: string;
 	  };
 
 type WorkspaceTimingsProps = {
 	defaultIsOpen?: boolean;
 	provisionerTimings: readonly ProvisionerTiming[] | undefined;
 	agentScriptTimings: readonly AgentScriptTiming[] | undefined;
+	agentConnectionTimings: readonly AgentConnectionTiming[] | undefined;
 };
 
 export const WorkspaceTimings: FC<WorkspaceTimingsProps> = ({
 	provisionerTimings = [],
 	agentScriptTimings = [],
+	agentConnectionTimings = [],
 	defaultIsOpen = false,
 }) => {
 	const [view, setView] = useState<TimingView>({ name: "default" });
-	const timings = [...provisionerTimings, ...agentScriptTimings];
+	const timings = [
+		...provisionerTimings,
+		...agentScriptTimings,
+		...agentConnectionTimings,
+	];
 	const [isOpen, setIsOpen] = useState(defaultIsOpen);
 	const isLoading = timings.length === 0;
 
 	const displayProvisioningTime = () => {
-		const totalRange = mergeTimeRanges(timings.map(extractRange));
+		const totalRange = mergeTimeRanges(timings.map(toTimeRange));
 		const totalDuration = calcDuration(totalRange);
 		return humanizeDuration(totalDuration);
 	};
 
+	const stages = () => {
+		const agentStageLabels = Array.from(
+			new Set(
+				agentConnectionTimings.map((t) => `agent (${t.workspace_agent_name})`),
+			),
+		);
+		return [
+			...provisioningStages,
+			...agentStageLabels.flatMap((a) => agentStages(a)),
+		];
+	};
+
```

Comment:
why put this in a function that doesn't take any arguments and is only called once? if it's a performance concern, we should probably be using `useMemo` instead

Reply:
That is true. I tried to improve readability by grouping related code to a function.